ipc/provider/src/manager/evm/manager.rs

```diff
@@ -100,32 +100,65 @@ impl TopDownCheckpointQuery for EthSubnetManager {
     async fn get_top_down_msgs(
         &self,
         subnet_id: &SubnetID,
-        start_epoch: ChainEpoch,
-        end_epoch: ChainEpoch,
-    ) -> Result<Vec<ipc_sdk::cross::CrossMsg>> {
-        self.top_down_msgs(subnet_id, start_epoch, end_epoch).await
+        epoch: ChainEpoch,
+    ) -> Result<TopDownQueryPayload<Vec<CrossMsg>>> {
+        let route = subnet_id_to_evm_addresses(subnet_id)?;
+        log::debug!("getting top down messages for route: {route:?}");
+
+        let subnet_id = gateway_getter_facet::SubnetID {
+            root: subnet_id.root_id(),
+            route,
+        };
+        let gateway_contract = gateway_getter_facet::GatewayGetterFacet::new(
+            self.ipc_contract_info.gateway_addr,
+            Arc::new(self.ipc_contract_info.provider.clone()),
+        );
+        let (raw_msgs, hash) = gateway_contract
+            .method::<_, (Vec<gateway_getter_facet::CrossMsg>, [u8; 32])>(
+                "getTopDownMsgs",
+                gateway_getter_facet::GetTopDownMsgsCall {
+                    subnet_id,
+                    block_number: U256::from(epoch),
+                },
+            )
```

Comment:
Why don't you use the statically typed ABI method?

Reply:
More of a legacy issue. The idea is we should query up till the target block, i.e. say if we want to get the messages up till block 100, but current head is 200. We still want to use the state at block 100 so that everyone has the same view. Back then this way was tested and works. It seems using the statically typed ABI could also support using block number state, but we have not tested it yet.

Reply:
Like [here](https://github.com/consensus-shipyard/fendermint/blob/main/fendermint/eth/api/examples/ethers.rs#L541-L548). I'm curious because if you did, you wouldn't have to declare the return type. The compiler would tell you if there is any mismatch too.

Reply:
Yeah, we can give it a try. Should work.